File: Temporada01/Episodio04 - No usar null/Java/Episodio04-Ejemplo-SinRomperEncapsulamiento/src/CustomerAgenda.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

public class CustomerAgenda {

	private List<Customer> customers;
	
	public CustomerAgenda() {
		customers = new ArrayList<Customer>();
	}
	
	public void add(Customer customerToAdd) {
		customers.add(customerToAdd);
	}

	public List<?> customersAt(String city) {
		return Arrays.asList((customers.stream()
			.filter(customer-> customer.addressDoIfNone(
				address-> address.isAt(city),
				()->false)))
			.toArray());
	}	

	public List<?> customersWithZipCode(String zipCode) {
		return Arrays.asList((customers.stream()
				.filter(customer-> customer.addressDoIfNone(
					address-> address.hasZipCode(zipCode),
					()->false)))
				.toArray());
	}
}

File: Temporada01/Episodio04 - No usar null/Ruby/Ejemplo-SinRomperEncapsulamiento/customer_agenda.rb
class CustomerAgenda
  def initialize
    @customers = []
  end

  def add(customer)
    @customers.push customer
  end

  def customers_at(city)
    @customers.select do |customer|
      customer.is_at(city)
    end
  end

  def customers_with_zip_code(zip_code)
    @customers.select do |customer|
      customer.has_zip_code zip_code
    end
  end
end